providers/dns/nicru/nicru.go

```diff
@@ -0,0 +1,329 @@
+	fqdn, value, ttl := acme.DNS01Record(domain, keyAuth)
+	if ttl > 60 {
+		ttl = 60
+	}
+
+	authZone, err := acme.FindZoneByFqdn(fqdn, acme.RecursiveNameservers)
+	if err != nil {
+		return fmt.Errorf("Could not determine zone for domain: '%s'. %s", domain, err)
+	}
+
+	authZone = acme.UnFqdn(authZone)
+
+	// http PUT https://api.nic.ru/dns-master/services/${NICRU_SERVICE_NAME}/zones/${ZONE_NAME}/records Authorization:"Bearer $TOKEN" @txt-record.xml
+
+	name := extractRecordName(fqdn, authZone)
+
+	reqTxt := &nicRequest{Records: []nicRecord{
+		{
+			Name:      name,
+			TTL:       ttl,
+			Type:      "TXT",
+			TxtString: value,
+		},
+	}}
+	reqTxtBody, err := xml.Marshal(reqTxt)
+	if err != nil {
+		return err
+	}
+
+	reqURL := fmt.Sprintf("%s/dns-master/services/%s/zones/%s/records",
+		d.baseURL, d.serviceName, authZone)
+
+	req, err := http.NewRequest("PUT", reqURL, bytes.NewReader(reqTxtBody))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/xml")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", d.auth.AccessToken))
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("TXT creation failed, code=%d with unreadable body (%s)", resp.StatusCode, err)
+		}
+
+		return fmt.Errorf("TXT creation failed, code=%d, body: %s", resp.StatusCode, body)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("TXT creation success, code=%d with unreadable body. error: %s", resp.StatusCode, err)
+	}
+
+	// Everything looks good, but try to decode response (should not fail in success case)
+	// MAYBE: keep record from ID later to delete the record
+	var respData nicResponse
+	err = xml.Unmarshal(body, &respData)
+	if err != nil {
+		return err
+	}
+
+	return d.commitChanges(authZone)
+}
+
+func extractRecordName(fqdn, domain string) string {
+	name := acme.UnFqdn(fqdn)
+	if idx := strings.Index(name, "."+domain); idx != -1 {
+		return name[:idx]
+	}
+	return name
+}
+
+// CleanUp removes the TXT records matching the specified parameters
+func (d *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	fqdn, _, _ := acme.DNS01Record(domain, keyAuth)
+
+	authZone, err := acme.FindZoneByFqdn(fqdn, acme.RecursiveNameservers)
+	if err != nil {
+		return fmt.Errorf("Could not determine zone for domain: '%s'. %s", domain, err)
+	}
+	authZone = acme.UnFqdn(authZone)
+
+	// find records
+	reqURL := fmt.Sprintf("%s/dns-master/services/%s/zones/%s/records",
+		d.baseURL, d.serviceName, authZone)
+	req, err := http.NewRequest("GET", reqURL, nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", d.auth.AccessToken))
+
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		body, err := ioutil.ReadAll(resp.Body)
```

Comment:
This could be moved outside the `if` statement because you do the same thing below too.